src/server/rdb_load.cc

```diff
@@ -2556,63 +2619,83 @@ void RdbLoader::ResizeDb(size_t key_num, size_t expire_num) {
   // load with different number of shards which makes database resizing unfeasible.
 }
 
+// Loads the next key/val pair.
+//
+// Huge objects may be loaded in parts, where only a subset of elements are
+// loaded at a time. This reduces the memory required to load huge objects and
+// prevents LoadItemsBuffer blocking. (Note so far only SETs support partial
+// loading).
 error_code RdbLoader::LoadKeyValPair(int type, ObjSettings* settings) {
-  // We return the item in LoadItemsBuffer.
-  Item* item = item_queue_.Pop();
+  std::string key;
+  SET_OR_RETURN(ReadKey(), key);
 
-  if (item == nullptr) {
-    item = new Item;
-  }
-  auto cleanup = absl::Cleanup([item] { delete item; });
+  do {
+    // If there is a cached Item in the free pool, take it, otherwise allocate
+    // a new Item.
+    Item* item = item_queue_.Pop();
+    if (item == nullptr) {
+      item = new Item;
+    }
+    // Delete the item if we fail to load the key/val pair.
+    auto cleanup = absl::Cleanup([item] { delete item; });
 
-  // Read key
-  SET_OR_RETURN(ReadKey(), item->key);
+    item->append = pending_read_.remaining > 0;
 
-  // Read value
-  error_code ec = ReadObj(type, &item->val);
-  if (ec) {
-    VLOG(1) << "ReadObj error " << ec << " for key " << item->key;
-    return ec;
-  }
+    error_code ec = ReadObj(type, &item->val);
+    if (ec) {
+      VLOG(1) << "ReadObj error " << ec << " for key " << item->key;
+      return ec;
+    }
 
-  if (!load_unowned_slots_ && cluster::IsClusterEnabled()) {
-    const cluster::ClusterConfig* cluster_config = cluster::ClusterFamily::cluster_config();
-    if (cluster_config != nullptr && !cluster_config->IsMySlot(item->key)) {
-      return kOk;  // Ignoring item
+    if (pending_read_.remaining > 0) {
```

Comment:
i would use `item->append` condition here and then would group the 3 ifs here, at line 2657 and 2673 accordingly.

Reply:
this is checking whether this is the last read in the object (which we can't infer from `append` which says whether it's the first read in the object)

> group the 3 ifs here, at line 2657 and 2673 accordingly.

i got that logic wrong (we need to check whether to discard on each read) - i've refactored